urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/core/Loginer.java

```diff
@@ -0,0 +1,226 @@
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Random;
+
+import javax.crypto.Cipher;
+import javax.crypto.CipherInputStream;
+import javax.crypto.CipherOutputStream;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import javax.crypto.spec.PBEParameterSpec;
+
+import com.kokakiwi.mclauncher.MCLauncher;
+
+public class Loginer
+{
+    private final MCLauncher main;
+    private Login            lastLogin      = new Login();
+    
+    private String           storedUsername = null;
+    private String           storedPassword = null;
+    
+    public Loginer(MCLauncher main)
+    {
+        this.main = main;
+    }
+    
+    public void init()
+    {
+        try
+        {
+            readLogin();
+        }
+        catch (final Exception e)
+        {
+            e.printStackTrace();
+        }
+    }
+    
+    // Actions
+    
+    public String doLogin(String username, String password, boolean rememberMe)
+    {
+        String result = null;
+        
+        final String url = main.getConfig().getString("login.url");
+        
+        final Map<String, String> params = new LinkedHashMap<String, String>();
+        params.put("user", username);
+        params.put("password", password);
+        params.put("version",
+                main.getConfig().getString("login.launcherVersion"));
+        
+        final String request = main.getApi().postUrl(url, params);
+        
+        if (request.contains(":"))
+        {
+            final String[] parts = request.split(":");
+            lastLogin = new Login();
+            lastLogin.setTimestamp(Long.parseLong(parts[0]));
+            lastLogin.setDownloadTicket(parts[1]);
+            lastLogin.setUsername(parts[2]);
+            lastLogin.setSessionId(parts[3]);
+        }
+        else
+        {
+            result = request;
+        }
+        
+        if (result == null)
+        {
+            if (rememberMe)
+            {
+                try
+                {
+                    storedUsername = username;
+                    storedPassword = password;
+                    storeLogin(username, password);
+                }
+                catch (final Exception e)
+                {
+                    e.printStackTrace();
+                }
+            }
+            else
+            {
+                deleteLogin();
+            }
+        }
+        
+        return result;
+    }
+    
+    // Utils
+    
+    public void storeLogin(String username, String password) throws Exception
+    {
+        final File file = new File(main.getApi().getMinecraftDirectory(),
+                "lastlogin");
+        
+        final Cipher cipher = getCipher(Cipher.ENCRYPT_MODE, main.getConfig()
+                .getString("login.encryptionKey"));
+        final DataOutputStream out = new DataOutputStream(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 110, 111, 112, 113. The resource is closed at line 114. There are other execution paths that do not contain closure statements, for example, when DataOutputStream.writeUTF throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).